src/main/kotlin/com/fasterxml/jackson/module/kotlin/Converters.kt

```diff
@@ -16,6 +18,18 @@ internal class SequenceToIteratorConverter(private val input: JavaType) : StdCon
         ?: typeFactory.constructType(Iterator::class.java)
 }
 
+internal object KotlinToJavaDurationConverter : StdConverter<KotlinDuration, JavaDuration>() {
+    override fun convert(value: KotlinDuration): JavaDuration = JavaDuration.parse(value.toIsoString())
+}
+
+// this class is needed as workaround for deserialization
+// data classes with kotlin.time.Duration field which is a value class
+//
+// @see DurationTests.`should deserialize Kotlin duration inside data class`
+object JavaToKotlinDurationConverter : StdConverter<JavaDuration, KotlinDuration>() {
```

Comment:
Is it possible to implement a `findDeserializationConverter` that returns a `Converter` for an `AnnotatedParameter`?
I think it can be implemented by just adding `Converter`.

https://github.com/ProjectMapK/jackson-module-kogera/blob/ca6bfe5a1689ebef4dff795ae21ccbe7ca0248df/src/main/kotlin/io/github/projectmapk/jackson/module/kogera/annotation_introspector/KotlinFallbackAnnotationIntrospector.kt#L92

Reply:
Unfortunately we can't do that without changes in Jackson core libs. 

**tl;dr;** Jackson extracts parameter of type `long` and `@JsonDeserialization` annotation is only source of its real type (in Java-reflection land).

### Why it happens

If we check ctors of tested class:
```
Meeting::class.constructors.forEach { println("kt ctor: $it") }
Meeting::class.java.declaredConstructors.forEach { println("java declared ctor: $it ${it.isSynthetic}") }
```

We will got:
```
kt ctor: fun <init>(java.time.Instant, kotlin.time.Duration): DurationTests.Meeting

java declared ctor: private DurationTests$Meeting(java.time.Instant,long) false
java declared ctor: public DurationTests$Meeting(java.time.Instant,long,kotlin.jvm.internal.DefaultConstructorMarker) true
```

Jackson looks only for Java ctors and **removed synthetic ctors** from lookup in https://github.com/FasterXML/jackson-databind/issues/1005 hence we will see only:
```
DurationTests$Meeting(java.time.Instant,long) 
```

### How we can restore original type

Theoretically, we can restore original types via Kotlin reflection **but this will work only for synthetic ctor**, following code:
```
Meeting::class.java.declaredConstructors.map { it.kotlinFunction }.forEach { println("restored kt ctor: $it") }
```
will return:
```
restored kt ctor: null
restored kt ctor: fun <init>(java.time.Instant, kotlin.time.Duration): DurationTests.Meeting
```

If we could collect that synthetic constructors (or just check for one with `kotlin.jvm.internal.DefaultConstructorMarker` param) then we can remove need of explicit `@JsonCreator` and `@JsonDeserializer`.

That being said, with current Jackson code we can't customize collection of constructors.

Reply:
Since annotations are applied to the type after conversion in serialization, I mistakenly thought that this was also true for deserialization.
If this is the case, I think it is sufficient to implement only the deserializer, so I am very sorry, but please remove `JavaToKotlinDurationConverter`.

Reply:
> If this is the case, I think it is sufficient to implement only the deserializer [...]

You mean this?
https://github.com/FasterXML/jackson-module-kotlin/pull/689/files#diff-0fe95cd6b35f4ae8bd509a735e78af8cd195263fdeeb9de7fdb30e1be802bb21R107

This won't cut the bill (at least in current code) because type we see from a ctor definition is a `long` hence we fall in wrong `when`-branch. Order in `when` clause doesn't matter.

> so I am very sorry, but please remove `JavaToKotlinDurationConverter`.

If I remove converter I need to remove it from POJO in `"should deserialize Kotlin duration inside data class"` and test will fail:
```
com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `long` from String "PT1H30M": not a valid `long` value
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 44] (through reference chain: com.fasterxml.jackson.module.kotlin.test.DurationTests$Meeting["duration"])
```

Reply:
Sorry, I seem to be confused myself.
I have compiled a list of proposed fixes for deserialization, could you please check it out?

The following points have been improved

- The `KotlinDuration` argument can now be deserialized without annotation
- Simplified by using `DelegatingDeserializer`
- Replaced `Java` -> `Kotlin` conversion with stdlib function

https://github.com/kkurczewski/jackson-module-kotlin/compare/duration-support...k163377:duration-support-review?expand=1

Reply:
> The KotlinDuration argument can now be deserialized without annotation

Nice, thank you!

> Simplified by using DelegatingDeserializer

Now I understand what you mean before, you wanted me to replace (remove) deserializer with converter like in case of serialization.

> Replaced Java -> Kotlin conversion with stdlib function

Ah, missed that one. Good catch.

Reply:
One minor thing, personally I would add guard condition to avoid `return@let` and use `when` at the end but I leave decision up to you.